### PR TITLE
Add document indexing section in deployer #7192

### PR DESCRIPTION
**Pull request by alhambrav**
### Ticket reference or full description of what's in the PR
Add document indexing section in deployer https://github.com/craftercms/craftercms/issues/7192

### CODE REVIEW

File: source/reference/modules/deployer/index.rst [hunk 1]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+        ...
+        binary:
+          # The list of binary file mime types that should be indexed
+          supportedMimeTypes:
+            - application/pdf
+            - application/msword
+            - application/vnd.openxmlformats-officedocument.wordprocessingml.document
+            - application/vnd.ms-excel
+            - application/vnd.ms-powerpoint
+            - application/vnd.openxmlformats-officedocument.presentationml.presentation
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:392]: How do you add to this list?

File: source/reference/modules/deployer/index.rst [hunk 2]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
   Changing or deleting a processed commit file could cause unchanged files to be indexed again and
   it should be done as a last resort in case of errors.
 
+^^^^^^^^^^^^^^^^^
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:357]: Throughout this PR, change mime to MIME. This also means it's MIME-type or MIME-types.

File: source/reference/modules/deployer/index.rst [hunk 3]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+Document Indexing
+^^^^^^^^^^^^^^^^^
+Crafter Deployer indexes documents via the following configured items:
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:360]: While the Deployer performs content indexing, we probably need to reference this higher up in the IA and make it clear(er). Let's discuss where else to point to this or include it. This is core to the system.

Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:360]: CrafterCMS indexes content items as follows

File: source/reference/modules/deployer/index.rst [hunk 4]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+
+- A full text index of any document that has a mimetype that matches the configured list of mimetypes.
+- Indexing of any remote document that matches the configured list of remote documents pattern
+- Indexing of jacketed documents with anything that matches the configured pattern.
+
+Note that indexing of documents in authoring and indexing of documents in delivery each have their own configuration.
+
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:367]: Additionally, CrafterCMS indexes non-text-based content as follows:
[add notes on how we index filenames and meta-data where possible]

File: source/reference/modules/deployer/index.rst [hunk 5]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
   Changing or deleting a processed commit file could cause unchanged files to be indexed again and
   it should be done as a last resort in case of errors.
 
+^^^^^^^^^^^^^^^^^
+Document Indexing
+^^^^^^^^^^^^^^^^^
+Crafter Deployer indexes documents via the following configured items:
+
+- A full text index of any document that has a mimetype that matches the configured list of mimetypes.
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:362]: - Full-text indexing of text-based documents like PDFs, MS Office, etc. See the list of supported MIME-types below.

File: source/reference/modules/deployer/index.rst [hunk 6]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+- Indexing of any remote document that matches the configured list of remote documents pattern
+- Indexing of jacketed documents with anything that matches the configured pattern.
+
+Note that indexing of documents in authoring and indexing of documents in delivery each have their own configuration.
+
+"""""""""
+Mimetypes
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:369]: Full-text Search MIME-Types

File: source/reference/modules/deployer/index.rst [hunk 7]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
   Changing or deleting a processed commit file could cause unchanged files to be indexed again and
   it should be done as a last resort in case of errors.
 
+^^^^^^^^^^^^^^^^^
+Document Indexing
+^^^^^^^^^^^^^^^^^
+Crafter Deployer indexes documents via the following configured items:
+
+- A full text index of any document that has a mimetype that matches the configured list of mimetypes.
+- Indexing of any remote document that matches the configured list of remote documents pattern
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:363]: - Full-text indexing of remote documents referenced by local content items, where the local item references match a document pattern [add a reference/link].

File: source/reference/modules/deployer/index.rst [hunk 8]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+- Indexing of jacketed documents with anything that matches the configured pattern.
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:364]: - Full-text indexing of text-based documents with a CrafterCMS content wrapper [reference what that is with a link and another link to the jacket example below].

File: source/reference/modules/deployer/index.rst [hunk 9]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:365]: - Indexing of filenames and meta-data for other content items. [Check what we do with binary files like .jpg, or similar where we have meta-data but not textual content and explain it here.]

File: source/reference/modules/deployer/index.rst [hunk 10]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+Note that indexing of documents in authoring and indexing of documents in delivery each have their own configuration.
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:366]: Use a _note_ directive.

Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:366]: Indexing is done differently in the authoring environment vs the delivery environment.

Authoring indexing is done to help content authors do their work and is controlled by CrafterCMS. The authoring search index is tuned to help authors and is not used by the project/site for delivery concerns.

On the other hand, delivery indexing is done to enable search and search-based features for the delivery project/site. This is configurable per project/site, and the index is tuned to help end-users use the project/site.

File: source/reference/modules/deployer/index.rst [hunk 11]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+
+"""""""""
+Mimetypes
+"""""""""
+The list of supported mimetypes determines what's considered a document that should be full-text indexed.
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:371]: The following is the default list of MIME-types with full-text-search indexing enabled.

File: source/reference/modules/deployer/index.rst [hunk 12]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+            - application/msword
+            - application/vnd.openxmlformats-officedocument.wordprocessingml.document
+            - application/vnd.ms-excel
+            - application/vnd.ms-powerpoint
+            - application/vnd.openxmlformats-officedocument.presentationml.presentation
+
+""""""""""""""""""""""""
+Remote Documents Pattern
+""""""""""""""""""""""""
+The ``remoteBinaryPathPatterns`` in the config determines what a remote document is via path pattern.
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:397]: What config? Name it up here before the example.

File: source/reference/modules/deployer/index.rst [hunk 13]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+          remoteBinaryPathPatterns: &remoteBinaryPathPatterns
+            # HTTP/HTTPS URLs are only indexed if they contain the protocol (http:// or https://). Protocol relative
+            # URLs (like //mydoc.pdf) are not supported since the protocol is unknown to the back-end indexer.
+            - ^(http:|https:)//.+$
+            - ^/remote-assets/.+$
+
+""""""""""""""
+Jacket Pattern
+""""""""""""""
+The ``metadataPathPatterns`` in the configuration determines if a document should be indexed with the metadata of the
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:422]: define "configuration" filename (like above)

File: source/reference/modules/deployer/index.rst [hunk 14]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+
+""""""""""""""""
+Indexing Failure
+""""""""""""""""
+When a document cannot be indexed:
+
+- The Deployer skips the document and leaves an error in the logs.
+- The document is not re-processed unless it is updated/re-deployed or the re-process API is called.
+
+The default behavior when a document cannot be indexed is that the Deployer logs the error and moves on. Process commits
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:450]: This repeats the bullets above. We should pick one.

File: source/reference/modules/deployer/index.rst [hunk 15]
```diff
@@ -355,6 +354,122 @@ the steps described in :ref:`debugging-deployer-issues`
...
+In these configurations, an additional processor is added to the end of the processor chain called the
+:ref:`Git Update Commit ID processor <deployer-git-update-commit-id-processor>` which only updates this ID on successful
+completion of the processor chain.
+
+Note that using this model can lead to A LOT of re-processing if there is no additional circuit breaker logging added
+for error conditions that do not resolve themselves.
+
+If the deployment as a whole cannot be completed due to a catastrophic exception, then all content including documents
+will be re-processed until the deployment succeeds. By default the Git Diff process is configured to update the processed
+commits regardless of success or failure. Some deployments set this to false and force the processor chain to be
```
Comment (sumerjabri) [inline source/reference/modules/deployer/index.rst:469]: This repeats the paragraphs above.

These paragraphs regarding the different configuration should be pulled into a single paragraph that describes this approach and details the pros and cons.

**Review by sumerjabri — CHANGES_REQUESTED** [review]

File: source/reference/modules/deployer/index.rst (hunk 14 above)
Comment (jmendeza) [inline source/reference/modules/deployer/index.rst:450]: Should it be "**Processed**" commits files ... ?

**Review by jmendeza — CHANGES_REQUESTED** [review]

File: source/reference/modules/deployer/index.rst (hunk 3 above)
Comment (sumerjabri) [reply source/reference/modules/deployer/index.rst:360]: This is where to discuss this: https://docs.craftercms.org/en/4.1/by-role/developer/common/search.html#content-search

File: source/reference/modules/deployer/index.rst (hunk 9 above)
Comment (sumerjabri) [reply source/reference/modules/deployer/index.rst:365]: The above is done for authoring. I don't think it's done for delivery. Check that, and if it's only delivery, remove it from this.

File: source/reference/modules/deployer/index.rst (hunk 10 above)
Comment (sumerjabri) [reply source/reference/modules/deployer/index.rst:366]: Let's drop references to authoring; so far, no one other than internal developers cares about this. We may reference this in authoring plugin development. @alhambrav create a ticket for documenting search mechanics for plugin development to be addressed in the future.

**Review by sumerjabri — APPROVED** [review]

**Pull request merged**